Wangscape/tilegen/CartesianPower.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "CartesianPowerIterator.h"
+#include <utility>
+
+namespace tilegen
+{
+
+template<typename InputIt>
+class CartesianPower
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag, 
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef CartesianPowerIterator<InputIt> const_iterator;
+
+    CartesianPower(InputIt first_, InputIt last_, size_t power_);
+    template<typename Container>
+    CartesianPower(const Container& container, size_t power_);
```

Comment:
> ... other methods' arguments should have "normal" naming, but if the're setting one of members, one should then find a new name or use this ugly `this->member` notation.

I don't think this is really a problem. Almost all functions which set a data member will set a `protected` or `private` data member, which will have an `m` prefix. In the rare case when a member method sets a `public` data member, `this->member = member` will not detract from readability.

When designing a class API, my priorities for readability and consistency are:
1. What the user will have to type to use the class: class name, `public` method names, `public` data members, (`public` dependent types).
2. What the user will have to read to use the class: method parameter names, template parameter names, `public` dependent types.
3. What the user will have read and write to make derived classes: `public` and `protected` members.
4. Private implementation details.

Since method parameter names rank below data member names, it's the method parameter names that will have to gain trailing underscores when necessary. I would be happy to require a trailing underscore in all method parameter names, but at the moment (AFAIK) it's not necessary because we have no class methods used to set a public data member to the value of a parameter.